### DISCUSSION TITLE
Support Anchor For Dropdown Menu

### DISCUSSION THREAD

**Discussion by BrianHung** [radix-ui/primitives#1418] [Ideas]
Popover menu has an anchor element to support positioning the menu separately from the trigger: https://www.radix-ui.com/docs/primitives/components/popover#anchor

It would be useful if dropdown menus could support a similar functionality.

**Comment by benoitgrelard** [comment]
Hi @BrianHung,

We added the functionality to `Popover` as they are quite a bit more generic.
With menus, it's hard to see why you wouldn't want it to drop from the trigger.
That being said, we are always interested to hear about your use-case so we can consider it.
What do you have in mind here?

✌️

**Reply by BrianHung** [reply]
Building a menu close to the dropdown menu from google sheets and also on top of a canvas element.

![image](https://user-images.githubusercontent.com/15721634/170101123-28bba72f-034f-4b70-97e9-02265c67997f.png)

Because its a canvas element, we only get a virtual element / viewport rectangle of the column header. As a workaround, I am using that header rectangle to position the menu trigger element and the menu content element.

However, the header rectangle changes on hover similar to google sheets: notice how the dropdown arrow moves to E while the content stays on F.

![image](https://user-images.githubusercontent.com/15721634/170101866-d59b4ab7-46c6-4865-bf35-24a4d96e9080.png)

Right now, I can achieve that behavior with popovers with anchor staying on column F and trigger moving to column E. But popovers lack support for nested menus and arrowkey item selection.

**Reply by sami616** [reply]
Interested to hear if anyone found a work around for this? @BrianHung @benoitgrelard  

I have a situation where right now I'm rendering the same dropdown hundreds of times in table rows which is causing some performance issues.

Ideally I would lift the dropdown and state up in my app then use buttons wrapped in anchors at the table cell level to position the menu against them respectivly.

**Comment by prtcl** [comment]
I know this is an old thread but I've come across this issue while working on a component. In my case, the trigger is a `Touchable`-like component which supports press and long press actions. For a regular press/tap I would like to trigger an action which opens a modal. For long press, I'm opening a controlled dropdown with a few options in the menu. It's not possible for me to `asChild` my custom trigger as there's collisions between which props Radix expects to spread on the button vs what's already being overridden to create the long press action. It looks like I should use a popover instead in my case, just wanted to report in on possible use cases for the `anchor` prop.

**Comment by markjaniczak** [comment]
An old thread but I'd also like this. 

There are times where I want the functionality of the [`menu`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/menu_role) component but I don't want it rendered within a popover adjacent to the trigger. Two examples:
- A responsive menu; on mobile I would want the menu to render in a dialog in the form of a drawer expanding from the bottom of the screen to provide a more native experience. Keyboard interactions provided by the `menu` component are [still needed even on mobile  to be truly accessible](https://accessibleweb.com/question-answer/why-do-mobile-apps-need-to-be-keyboard-accessible/)
- A trigger button within an anchoring element. For example, Notion's form builder displays a trigger button within each question but the dropdown menu is instead aligned with the card:
<img width="1065" alt="Screenshot 2025-03-26 at 10 53 17 pm" src="https://github.com/user-attachments/assets/a0352521-4bdb-4b79-b5f9-aa4a9b0f6b1c" />

While it is possible to use the popover anchor to achieve this kind of positioning, we don't have a separate menu component to place inside of that popover.

Some possible solutions:
- Extract the menu list functionality out to its own primitive so that it can be composed with other components
- Add a `DropdownMenu.Anchor` component similar to `Popover.Anchor`
- Expose more options from the floating UI package

**Comment by arthur-adriansens** [comment]
I also like this idea. I have a bunch of cards with a button to show more options, but I don't want to slow down my app with all these dropdown menu's.
The triggers are the buttons with the three dots.

<img width="291" height="296" alt="image" src="https://github.com/user-attachments/assets/1ce47cc6-68eb-48f5-8433-de681fd01768" />

**Comment by CodeSmith32** [comment]
There's currently a bug with DropdownMenuTriggers / PopoverTriggers that prevents them from working together, when the dropdown menu trigger also has a tooltip, e.g.:

```tsx
<DropdownMenu>
  <DropdownMenu.Trigger asChild>

    {/* <component boundary> */}

    <Tooltip.Provider>
      <Tooltip.Root>
        <Tooltip.Trigger>Menu</Tooltip.Trigger>
        <Tooltip.Content>Hello world</Tooltip.Content>
      </Tooltip.Root>
    </Tooltip.Provider>

    {/* </component boundary> */}

  </DropdownMenu.Trigger>

  <DropdownMenu.Content>
    ...
  </DropdownMenu.Content>
<DropdownMenu>
```

In this case, the tooltip overtakes the trigger element, and the dropdown menu is *not opened* when clicking the trigger. I think there's a fix that mixes these primitive elements together, but in my case, as the dropdown trigger, I'm trying to use a component that already pre-wraps its contents in a tooltip.

Now, this isn't a problem for Popovers because we can just take control with `open` / `onOpenChange` and wrap the tooltip in the anchor. Works great!

Apparently, we can't fix this easily with DropdownMenus, since we don't have anchors. Even wrapping the tooltip-wrapped trigger with `<DropdownMenu.Trigger>` still causes the menu to open up off the top of the screen. At this point, re-creating the menu with a Popover is easier.

+1

------

EDIT:

Ok, actually, I think I just set it up wrong. It works if the component is set up correctly, by forwarding a ref to a button wrapping the trigger. e.g.,

```tsx
const TooltipButton = fowardedRef(({ label, tooltip }, ref) => {
  return (
    <Tooltip.Provider>
      <Tooltip.Root>
        <Tooltip.Trigger>
          <button ref={ref}>
            {label}
          </button>
        </Tooltip.Trigger>
        <Tooltip.Content>
          {tooltip}
        </Tooltip.Content>
      </Tooltip.Root>
    </Tooltip.Provider>
  );
});
```

Note that, even with this setup, I still needed to use `open` / `onOpenChange`. But at least the menu trigger is anchored correctly.
